R/FactbaseWidget.R

```diff
@@ -0,0 +1,419 @@
+                    possible.data.mess <- c(possible.data.mess, 
+                                            paste0(v.names[j], " looks to contain a mix of numeric and text entries."))
+                prop.unique <- variable[["Summary"]]["Unique Entries"] / n.cases
+                if (prop.unique > 0.95 && prop.unique < 1 ) {
+                    possible.data.mess <- c(possible.data.mess,
+                                            paste0(v.names[j], " looks like an ID variable but its entries are not unique.")) 
+                }
+            }
+        }
+    }
+    list(data.integrity.issues = data.integrity.issues,
+         possible.data.mess = possible.data.mess)
+}
+
+
+FactBaseOutput <- function(input.data, data.summaries, output.type = "Data Summaries") {
+    if (output.type == "Data")
+        return (input.data)
+    
+    CreateFactBaseMetricSummary(data.summaries)   
+}
+
+CheckDataForFactBaseMetricUpload <- function(input.data, warn = TRUE) {
+    data.summaries <- DataSummaryForFactBase(input.data)
+    date.issues <- CheckFactBaseDateVariable(input.data[, "_When"])
+    data.issues <- IdentifyFactBaseDataIssues(data.summaries)
+    if (warn) {
+        all.issues <- c(date.issues, unlist(data.issues))
+        lapply(all.issues, FUN = function (x) warning(x))
+    }
+    attr(data.summaries, "date.issues") <- date.issues
+    attr(data.summaries, "data.issues") <- data.issues
+    attr(data.summaries, "data.errors") <- length(date.issues) > 0 || length(data.issues[["data.integrity.issues"]]) > 0
+    data.summaries
+}
+
+ParseFactBaseOption <- function(selection) {
+    if (is.null(selection))
+        return(NULL)
+    selection <- gsub(" ", "_", selection)
+    selection <- tolower(selection)
+    selection
+}
+
+ConditioanllyRenameVariables <- function(data, names) {
+    current.names <- colnames(data)
+    override.inds <- which(nzchar(names))
+    if (length(override.inds) == 0)
+        return(data)
+    
+    current.names[override.inds] <- names[override.inds]
+    colnames(data) <- current.names
+    data
+}
+
+CreateFactBaseMetricSummary <- function (x) {
+    require(flipFormat)
+    require(knitr)
+    addTable <-function(data) {
+        out <- knitr::kable(data, format = "html", col.names = "",
+                            table.attr = "class=\"cmd-table-one-stat\"")
+        ## change table headers to span multiple columns
+        out <- gsub("<th style=\"text-align:right;\">  </th>", "", out, fixed = TRUE)
+        out <- gsub("<th style=\"text-align:left;\">",
+                    "<th colspan = \"2\">", out, fixed = TRUE)
+        return(out)
+    }
+    
+    missing.by.variable <- vapply(x, FUN = function (x) x[["Missing"]], FUN.VALUE = numeric(1))
+    variable.summaries <- lapply(x, FUN = function (x) x[["Summary"]])
+    variable.types <- vapply(x, FUN = function(x) x[["Type"]], FUN.VALUE = character(1))
+    
+    
+    tfile <- flipFormat:::createTempFile()
```

Comment:
For building this as a package we should:
1. Use importFrom flipFormat createTempFile createCata addCss boxIframeless
2. Remove all the flipFormat::: and just use the bare function name which is now imported.

Reply:
Thanks, Chris, this mostly worked, but not for the `flipFormat:::` imports, because roxygen refuses to import non-exported members.  Can I export them, but mark them with `#' @noRd` so they are not documented, as recommended at https://devguide.ropensci.org/pkg_building.html?

Reply:
I've tried it at https://github.com/Displayr/flipFormat/pull/47, but it's failing because it's angry that the functions are not exported.  I could instead use `#' keyword internal`, I've heard...

Reply:
I think that might be the best option for now. Justin W may be able to offer an alternative when he is back. I don't think there is any harm in exporting those functions.